### DISCUSSION TITLE
Meeting 2 Agenda / Notes 1/18/2022

### DISCUSSION THREAD

**Discussion by robertlong** [omigroup/omi-scripting-group#12] [General]
# Agenda
- Adam will give a high level overview of WASM so we can get our bearings as a group talking about WASM specifications

**Comment by robertlong** [comment]
# Attendance

![image](https://user-images.githubusercontent.com/1753624/149977945-7d06cd61-5811-48b7-a14d-0838faa4b923.png)

# Notes

### WASM: Quick Overview by @AdamFrisby 

- What is WASM
  - Effectively a compiled program for the web
  - Evolved out of asm.js ecosystem
  - Designed as a potential successor to JavaScript on the web
- Benefits for the metaverse:
  - It's safe and secure, designed for untrusted uses
  - It's got effective resource limits (some via runtime)
  - It's mostly language agnostic
  - Runtime exist for just about every programming environment
- Neat Features
  - Memory is a byte array; serialize your runtime state and unpack it in the next environment!
  - Modules allow safe inter-process communication between other modules
  - Lots of mature tools such as WASMFiddle
- What's going to drive you crazy?
  - No standard ABI
  - No standard memory allocator
  - Effectively seven fixed with types
  - Strings (not in the spec)
- Embeddability
  - Unity
    - Wasmtime
    - Wasmer
  - Web
    - First class citizen
  - Other Projects
    - WASM Micro Runtime
    - Wasmer
    - Wasmtime
    - Etc.
- Quick Case Study
  - Sinespace
    - 128 Public Scripting Classes
    - 5- 20 Members in each (~1,500-2,000 methods in total)
  - Currently classes are exposed to both Lua and C# scripting runtimes already
  - WASM Support incoming soon, but requires a new way of thinking
- Adam gives an example of how it's implemented in Sinespace


- Post-presentation discussion
  - Likely easy to embed into Unreal as well since it is written in C++ and Rust/C++ bindings will be ready to use. For Unity, support was made easier because of C# bindings which you wouldn't need in C++
  - WASM bundle size may not be as important for our use-cases. We're shipping media where the script file size can be a rounding error.
  - Garbage collection spec has been proposed for WASM. Would like to know if this is predominantly a file-size thing of if there are performance benefits.
  - We may want to focus on a few languages for our own examples. We should ensure that it should be usable across many different types of languages though.
  - https://hacks.mozilla.org/2019/08/webassembly-interface-types/
  - https://github.com/webassembly/component-model/
  - WITX possible to be used for ABI definition? https://github.com/WebAssembly/WASI/tree/main/phases
  - We may want to focus on exposing a glTF scene graph to start and not bring security issues around scripts being able to modify objects outside it's own scene graph.

**Comment by Trensharo** [comment]
Some Notes taken from the WASM Review Meeting

**What is WASM?**
- Equivalent of a .exe
- Has a binary format
- Designed for small programs, but can do big programs
- Safe and Secure -  Does not allow attackers access to  your inner systems
- Resource limits built-in
- Language Agnostic
- Easy to send the script's memory as a byte array

**WASM Potential Issues**
- Interoperability standards don't exist

**Projects using WASM**
- Lunatic - Web Assembly Project
- OpenSim

**Bucket of Ideas**
- Global APIs like Raycasting
- GLTF format to start
- Backwards compatible wrappers for JavaScript
- Define Extensions as Modules
- Good Quick Start Guide

**To-Dos**
- Should work on Permission Management
- Scope Declaration
